File: backend-v2/frontend-v2/types/review.ts
/**
 * Review management types for BookedBarber V2
 * Based on backend schemas from schemas_new/review.py
 */

export enum ReviewPlatform {
  GOOGLE = "google",
  YELP = "yelp",
  FACEBOOK = "facebook",
  INSTAGRAM = "instagram",
  BOOKSY = "booksy",
  FRESHA = "fresha",
  STYLESEAT = "styleseat",
  OTHER = "other"
}

export enum ReviewSentiment {
  POSITIVE = "positive",
  NEUTRAL = "neutral",
  NEGATIVE = "negative",
  UNKNOWN = "unknown"
}

export enum ReviewResponseStatus {
  PENDING = "pending",
  SENT = "sent",
  FAILED = "failed",
  NOT_NEEDED = "not_needed",
  DRAFT = "draft"
}

export interface Review {
  id: number
  user_id: number
  platform: ReviewPlatform
  external_review_id: string
  business_id?: string
  reviewer_name?: string
  reviewer_photo_url?: string
  rating: number
  review_text?: string
  review_date: string
  review_url?: string
  
  // Sentiment analysis
  sentiment: ReviewSentiment
  sentiment_score?: number
  sentiment_confidence?: number
  sentiment_keywords: string[]
  
  // Response tracking
  response_status: ReviewResponseStatus
  response_text?: string
  response_date?: string
  response_author?: string
  auto_response_generated: boolean
  
  // Business intelligence
  keywords_mentioned: string[]
  services_mentioned: string[]
  competitor_mentions: string[]
  location_mentions: string[]
  
  // Quality metrics
  is_verified: boolean
  is_flagged: boolean
  flag_reason?: string
  is_helpful?: boolean
  helpful_count: number
  
  // Metadata
  last_synced_at?: string
  is_deleted_on_platform: boolean
  sync_errors: Record<string, any>[]
  created_at: string
  updated_at: string
  
  // Computed properties
  is_positive: boolean
  is_negative: boolean
  needs_response: boolean
  can_respond: boolean
}

export interface ReviewResponse {
  id: number
  review_id: number
  user_id: number
  response_text: string
  response_type: string
  template_id?: string
  is_draft: boolean
  is_sent: boolean
  sent_at?: string
  platform_response_id?: string
  
  // SEO and optimization
  keywords_used: string[]
  cta_included: boolean
  business_name_mentioned: boolean
  
  // Performance tracking
  view_count: number
  helpful_votes: number
  character_count: number
  
  // A/B testing
  variant_id?: string
  test_group?: string
  
  // Error tracking
  send_errors: Record<string, any>[]
  last_error?: string
  
  // Metadata
  created_at: string
  updated_at: string
  
  // Computed properties
  is_ready_to_send: boolean
}

export interface ReviewTemplate {
  id: number
  user_id: number
  name: string
  description?: string
  category: string
  platform?: ReviewPlatform
  template_text: string
  placeholders: string[]
  
  // Conditions
  min_rating?: number
  max_rating?: number
  keywords_trigger: string[]
  sentiment_trigger?: ReviewSentiment
  
  // SEO
  seo_keywords: string[]
  include_business_name: boolean
  include_cta: boolean
  cta_text?: string
  
  // Usage tracking
  use_count: number
  success_rate: number
  avg_response_time: number
  
  // Settings
  is_active: boolean
  is_default: boolean
  priority: number
  
  // Metadata
  created_at: string
  updated_at: string
  last_used_at?: string
}

export interface ReviewAnalytics {
  total_reviews: number
  average_rating: number
  rating_distribution: Record<number, number>
  
  // Platform breakdown
  platform_breakdown: Record<string, Record<string, any>>
  
  // Sentiment analysis
  sentiment_breakdown: Record<string, number>
  positive_percentage: number
  negative_percentage: number
  
  // Response metrics
  response_rate: number
  avg_response_time_hours: number
  auto_response_percentage: number
  
  // Time-based metrics
  reviews_this_month: number
  reviews_last_month: number
  month_over_month_change: number
  
  // Quality metrics
  verified_reviews_count: number
  flagged_reviews_count: number
  helpful_reviews_count: number
  
  // SEO insights
  top_keywords: Array<{ keyword: string; count: number; sentiment: string }>
  services_mentioned: Array<{ service: string; count: number; rating: number }>
  competitor_mentions: string[]
}

export interface ReviewFilters {
  platform?: ReviewPlatform
  sentiment?: ReviewSentiment
  response_status?: ReviewResponseStatus
  min_rating?: number
  max_rating?: number
  start_date?: string
  end_date?: string
  is_flagged?: boolean
  is_verified?: boolean
  has_response?: boolean
  search_query?: string
  business_id?: string
}

export interface ReviewSyncRequest {
  platform: ReviewPlatform
  business_id?: string
  force_full_sync?: boolean
  sync_responses?: boolean
  date_range_days?: number
}

export interface ReviewSyncResponse {
  success: boolean
  message: string
  synced_at: string
  platform: ReviewPlatform
  business_id?: string
  reviews_synced: number
  new_reviews: number
  updated_reviews: number
  errors_count: number
  total_reviews_after_sync: number
  average_rating_after_sync: number
  errors: string[]
}

export interface GMBLocation {
  location_id: string
  name: string
  address: string
  phone?: string
  website?: string
  category?: string
  hours?: Record<string, any>
  rating?: number
  review_count?: number
  is_verified: boolean
  is_published: boolean
}

export interface ReviewResponseCreate {
  response_text: string
  template_id?: string
  keywords_used?: string[]
  cta_included?: boolean
  business_name_mentioned?: boolean
  variant_id?: string
  test_group?: string
}

export interface ReviewResponseUpdate {
  response_text?: string
  is_draft?: boolean
  keywords_used?: string[]
  cta_included?: boolean
  business_name_mentioned?: boolean
}

export interface ReviewTemplateCreate {
  name: string
  description?: string
  category: string
  platform?: ReviewPlatform
  template_text: string
  placeholders?: string[]
  min_rating?: number
  max_rating?: number
  keywords_trigger?: string[]
  sentiment_trigger?: ReviewSentiment
  seo_keywords?: string[]
  include_business_name?: boolean
  include_cta?: boolean
  cta_text?: string
  is_active?: boolean
  is_default?: boolean
  priority?: number
}

export interface ReviewTemplateUpdate {
  name?: string
  description?: string
  category?: string
  platform?: ReviewPlatform
  template_text?: string
  placeholders?: string[]
  min_rating?: number
  max_rating?: number
  keywords_trigger?: string[]
  sentiment_trigger?: ReviewSentiment
  seo_keywords?: string[]
  include_business_name?: boolean
  include_cta?: boolean
  cta_text?: string
  is_active?: boolean
  is_default?: boolean
  priority?: number
}

export interface BulkResponseRequest {
  review_ids: number[]
  template_id?: number
  auto_send?: boolean
  business_name?: string
}

export interface BulkResponseResult {
  review_id: number
  success: boolean
  response_id?: number
  error?: string
  response_text?: string
}

export interface BulkResponseResponse {
  success: boolean
  message: string
  total_processed: number
  successful_responses: number
  failed_responses: number
  results: BulkResponseResult[]
}

export interface AutoResponseConfig {
  enabled: boolean
  respond_to_positive: boolean
  respond_to_neutral: boolean
  respond_to_negative: boolean
  min_rating_threshold: number
  max_rating_threshold: number
  delay_hours: number
  platforms: ReviewPlatform[]
  business_hours_only: boolean
  weekend_responses: boolean
  template_selection_strategy: string
}

export interface AutoResponseStats {
  total_auto_responses: number
  auto_responses_today: number
  auto_responses_this_week: number
  auto_responses_this_month: number
  success_rate: number
  average_response_time_hours: number
  most_used_template?: string
  platform_breakdown: Record<string, number>
}

// API Response wrapper types
export interface ReviewsListResponse {
  reviews: Review[]
  total: number
  skip: number
  limit: number
  has_more: boolean
}

export interface TemplateGenerateRequest {
  template_id: number
  business_name?: string
  custom_placeholders?: Record<string, string>
}

export interface TemplateGenerateResponse {
  success: boolean
  response_text: string
  template_used: string
  placeholders_replaced: Record<string, string>
  character_count: number
  seo_keywords_included: string[]
}

File: backend-v2/schemas_new/review.py
"""
Pydantic schemas for review management and Google My Business integration.
Handles validation for reviews, responses, templates, and analytics.
"""

from pydantic import BaseModel, Field, field_validator
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class ReviewPlatform(str, Enum):
    """Supported review platforms"""
    GOOGLE = "google"
    YELP = "yelp"
    FACEBOOK = "facebook"
    INSTAGRAM = "instagram"
    BOOKSY = "booksy"
    FRESHA = "fresha"
    STYLESEAT = "styleseat"
    OTHER = "other"


class ReviewSentiment(str, Enum):
    """Review sentiment analysis results"""
    POSITIVE = "positive"
    NEUTRAL = "neutral"
    NEGATIVE = "negative"
    UNKNOWN = "unknown"


class ReviewResponseStatus(str, Enum):
    """Status of review response"""
    PENDING = "pending"
    SENT = "sent"
    FAILED = "failed"
    NOT_NEEDED = "not_needed"
    DRAFT = "draft"


# Review Schemas
class ReviewBase(BaseModel):
    """Base review schema"""
    platform: ReviewPlatform
    external_review_id: str = Field(..., description="Platform-specific review ID")
    business_id: Optional[str] = Field(None, description="External business ID")
    reviewer_name: Optional[str] = Field(None, max_length=255)
    rating: float = Field(..., ge=1, le=5, description="1-5 star rating")
    review_text: Optional[str] = Field(None, description="Review content")
    review_date: datetime
    review_url: Optional[str] = Field(None, description="Direct link to review")


class ReviewCreate(ReviewBase):
    """Schema for creating a new review"""
    reviewer_photo_url: Optional[str] = None
    platform_data: Optional[Dict[str, Any]] = Field(default_factory=dict)
    is_verified: Optional[bool] = False
    
    @field_validator('external_review_id')
    @classmethod
    def validate_external_id(cls, v):
        if not v.strip():
            raise ValueError('External review ID cannot be empty')
        return v.strip()


class ReviewUpdate(BaseModel):
    """Schema for updating a review"""
    reviewer_name: Optional[str] = Field(None, max_length=255)
    rating: Optional[float] = Field(None, ge=1, le=5)
    review_text: Optional[str] = None
    review_url: Optional[str] = None
    sentiment: Optional[ReviewSentiment] = None
    sentiment_score: Optional[float] = Field(None, ge=-1, le=1)
    sentiment_confidence: Optional[float] = Field(None, ge=0, le=1)
    is_flagged: Optional[bool] = None
    flag_reason: Optional[str] = Field(None, max_length=500)
    platform_data: Optional[Dict[str, Any]] = None


class ReviewResponse(ReviewBase):
    """Schema for review responses"""
    id: int
    user_id: int
    business_id: Optional[str] = None
    reviewer_photo_url: Optional[str] = None
    
    # Sentiment analysis
    sentiment: ReviewSentiment
    sentiment_score: Optional[float] = None
    sentiment_confidence: Optional[float] = None
    sentiment_keywords: List[str] = Field(default_factory=list)
    
    # Response tracking
    response_status: ReviewResponseStatus
    response_text: Optional[str] = None
    response_date: Optional[datetime] = None
    response_author: Optional[str] = None
    auto_response_generated: bool = False
    
    # Business intelligence
    keywords_mentioned: List[str] = Field(default_factory=list)
    services_mentioned: List[str] = Field(default_factory=list)
    competitor_mentions: List[str] = Field(default_factory=list)
    location_mentions: List[str] = Field(default_factory=list)
    
    # Quality metrics
    is_verified: bool = False
    is_flagged: bool = False
    flag_reason: Optional[str] = None
    is_helpful: Optional[bool] = None
    helpful_count: int = 0
    
    # Metadata
    last_synced_at: Optional[datetime] = None
    is_deleted_on_platform: bool = False
    sync_errors: List[Dict[str, Any]] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    
    # Computed properties
    is_positive: bool = Field(default=False, description="Whether review is positive")
    is_negative: bool = Field(default=False, description="Whether review is negative")
    needs_response: bool = Field(default=False, description="Whether review needs a response")
    can_respond: bool = Field(default=False, description="Whether we can respond to this review")
    
    class Config:
        from_attributes = True


# Review Response Schemas
class ReviewResponseBase(BaseModel):
    """Base review response schema"""
    response_text: str = Field(..., min_length=1, max_length=4096)
    response_type: str = Field(default="custom", description="Type of response")
    template_id: Optional[str] = Field(None, description="Template used for response")


class ReviewResponseCreate(ReviewResponseBase):
    """Schema for creating a review response"""
    keywords_used: Optional[List[str]] = Field(default_factory=list)
    cta_included: Optional[bool] = False
    business_name_mentioned: Optional[bool] = False
    variant_id: Optional[str] = Field(None, description="A/B test variant")
    test_group: Optional[str] = Field(None, description="Test group")
    
    @field_validator('response_text')
    @classmethod
    def validate_response_text(cls, v):
        v = v.strip()
        if not v:
            raise ValueError('Response text cannot be empty')
        if len(v) > 4096:
            raise ValueError('Response text too long (max 4096 characters)')
        return v


class ReviewResponseUpdate(BaseModel):
    """Schema for updating a review response"""
    response_text: Optional[str] = Field(None, min_length=1, max_length=4096)
    is_draft: Optional[bool] = None
    keywords_used: Optional[List[str]] = None
    cta_included: Optional[bool] = None
    business_name_mentioned: Optional[bool] = None


class ReviewResponseSchema(ReviewResponseBase):
    """Schema for review response display"""
    id: int
    review_id: int
    user_id: int
    is_draft: bool = True
    is_sent: bool = False
    sent_at: Optional[datetime] = None
    platform_response_id: Optional[str] = None
    
    # SEO and optimization
    keywords_used: List[str] = Field(default_factory=list)
    cta_included: bool = False
    business_name_mentioned: bool = False
    
    # Performance tracking
    view_count: int = 0
    helpful_votes: int = 0
    character_count: int = 0
    
    # A/B testing
    variant_id: Optional[str] = None
    test_group: Optional[str] = None
    
    # Error tracking
    send_errors: List[Dict[str, Any]] = Field(default_factory=list)
    last_error: Optional[str] = None
    
    # Metadata
    created_at: datetime
    updated_at: datetime
    
    # Computed properties
    is_ready_to_send: bool = Field(default=False, description="Whether response is ready to send")
    
    class Config:
        from_attributes = True


# Review Template Schemas
class ReviewTemplateBase(BaseModel):
    """Base review template schema"""
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    category: str = Field(..., description="positive, negative, neutral")
    platform: Optional[ReviewPlatform] = None
    template_text: str = Field(..., min_length=1)


class ReviewTemplateCreate(ReviewTemplateBase):
    """Schema for creating a review template"""
    placeholders: Optional[List[str]] = Field(default_factory=list)
    min_rating: Optional[float] = Field(None, ge=1, le=5)
    max_rating: Optional[float] = Field(None, ge=1, le=5)
    keywords_trigger: Optional[List[str]] = Field(default_factory=list)
    sentiment_trigger: Optional[ReviewSentiment] = None
    seo_keywords: Optional[List[str]] = Field(default_factory=list)
    include_business_name: Optional[bool] = True
    include_cta: Optional[bool] = True
    cta_text: Optional[str] = Field(None, max_length=500)
    is_active: Optional[bool] = True
    is_default: Optional[bool] = False
    priority: Optional[int] = 0
    
    @field_validator('category')
    @classmethod
    def validate_category(cls, v):
        valid_categories = ['positive', 'negative', 'neutral']
        if v.lower() not in valid_categories:
            raise ValueError(f'Category must be one of: {", ".join(valid_categories)}')
        return v.lower()
    
    @field_validator('max_rating')
    @classmethod
    def validate_rating_range(cls, v, info):
        if v is not None and info.data and 'min_rating' in info.data and info.data['min_rating'] is not None:
            if v < info.data['min_rating']:
                raise ValueError('max_rating must be greater than or equal to min_rating')
        return v


class ReviewTemplateUpdate(BaseModel):
    """Schema for updating a review template"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    category: Optional[str] = None
    platform: Optional[ReviewPlatform] = None
    template_text: Optional[str] = Field(None, min_length=1)
    placeholders: Optional[List[str]] = None
    min_rating: Optional[float] = Field(None, ge=1, le=5)
    max_rating: Optional[float] = Field(None, ge=1, le=5)
    keywords_trigger: Optional[List[str]] = None
    sentiment_trigger: Optional[ReviewSentiment] = None
    seo_keywords: Optional[List[str]] = None
    include_business_name: Optional[bool] = None
    include_cta: Optional[bool] = None
    cta_text: Optional[str] = Field(None, max_length=500)
    is_active: Optional[bool] = None
    is_default: Optional[bool] = None
    priority: Optional[int] = None


class ReviewTemplateSchema(ReviewTemplateBase):
    """Schema for review template display"""
    id: int
    user_id: int
    placeholders: List[str] = Field(default_factory=list)
    
    # Conditions
    min_rating: Optional[float] = None
    max_rating: Optional[float] = None
    keywords_trigger: List[str] = Field(default_factory=list)
    sentiment_trigger: Optional[ReviewSentiment] = None
    
    # SEO
    seo_keywords: List[str] = Field(default_factory=list)
    include_business_name: bool = True
    include_cta: bool = True
    cta_text: Optional[str] = None
    
    # Usage tracking
    use_count: int = 0
    success_rate: float = 0.0
    avg_response_time: float = 0.0
    
    # Settings
    is_active: bool = True
    is_default: bool = False
    priority: int = 0
    
    # Metadata
    created_at: datetime
    updated_at: datetime
    last_used_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


# Analytics Schemas
class ReviewAnalytics(BaseModel):
    """Review analytics summary"""
    total_reviews: int = 0
    average_rating: float = 0.0
    rating_distribution: Dict[int, int] = Field(default_factory=dict)  # {1: count, 2: count, ...}
    
    # Platform breakdown
    platform_breakdown: Dict[str, Dict[str, Any]] = Field(default_factory=dict)
    
    # Sentiment analysis
    sentiment_breakdown: Dict[str, int] = Field(default_factory=dict)
    positive_percentage: float = 0.0
    negative_percentage: float = 0.0
    
    # Response metrics
    response_rate: float = 0.0
    avg_response_time_hours: float = 0.0
    auto_response_percentage: float = 0.0
    
    # Time-based metrics
    reviews_this_month: int = 0
    reviews_last_month: int = 0
    month_over_month_change: float = 0.0
    
    # Quality metrics
    verified_reviews_count: int = 0
    flagged_reviews_count: int = 0
    helpful_reviews_count: int = 0
    
    # SEO insights
    top_keywords: List[Dict[str, Any]] = Field(default_factory=list)
    services_mentioned: List[Dict[str, Any]] = Field(default_factory=list)
    competitor_mentions: List[str] = Field(default_factory=list)


class ReviewFilters(BaseModel):
    """Filters for review queries"""
    platform: Optional[ReviewPlatform] = None
    sentiment: Optional[ReviewSentiment] = None
    response_status: Optional[ReviewResponseStatus] = None
    min_rating: Optional[float] = Field(None, ge=1, le=5)
    max_rating: Optional[float] = Field(None, ge=1, le=5)
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    is_flagged: Optional[bool] = None
    is_verified: Optional[bool] = None
    has_response: Optional[bool] = None
    search_query: Optional[str] = Field(None, max_length=255)
    business_id: Optional[str] = None


class ReviewSyncRequest(BaseModel):
    """Request to sync reviews from platform"""
    platform: ReviewPlatform
    business_id: Optional[str] = None
    force_full_sync: Optional[bool] = False
    sync_responses: Optional[bool] = True
    date_range_days: Optional[int] = Field(30, ge=1, le=365, description="Number of days to sync")


class ReviewSyncResponse(BaseModel):
    """Response after review sync operation"""
    success: bool
    message: str
    synced_at: datetime
    platform: ReviewPlatform
    business_id: Optional[str] = None
    
    # Sync results
    reviews_synced: int = 0
    new_reviews: int = 0
    updated_reviews: int = 0
    errors_count: int = 0
    
    # Summary
    total_reviews_after_sync: int = 0
    average_rating_after_sync: float = 0.0
    
    errors: List[str] = Field(default_factory=list)


# GMB-specific schemas
class GMBLocation(BaseModel):
    """Google My Business location information"""
    location_id: str = Field(..., description="GMB location ID")
    name: str
    address: str
    phone: Optional[str] = None
    website: Optional[str] = None
    category: Optional[str] = None
    hours: Optional[Dict[str, Any]] = None
    rating: Optional[float] = None
    review_count: Optional[int] = None
    is_verified: bool = False
    is_published: bool = False


class GMBAuthRequest(BaseModel):
    """Request to initiate GMB OAuth flow"""
    redirect_uri: Optional[str] = None
    scopes: Optional[List[str]] = Field(default_factory=lambda: [
        "https://www.googleapis.com/auth/business.manage"
    ])


class GMBAuthResponse(BaseModel):
    """Response after GMB OAuth completion"""
    success: bool
    message: str
    auth_url: Optional[str] = None
    integration_id: Optional[int] = None


class ReviewTemplateGenerateRequest(BaseModel):
    """Request to generate response using template"""
    template_id: int
    business_name: Optional[str] = None
    custom_placeholders: Optional[Dict[str, str]] = Field(default_factory=dict)


class ReviewTemplateGenerateResponse(BaseModel):
    """Generated response from template"""
    success: bool
    response_text: str
    template_used: str
    placeholders_replaced: Dict[str, str] = Field(default_factory=dict)
    character_count: int
    seo_keywords_included: List[str] = Field(default_factory=list)


class BulkResponseRequest(BaseModel):
    """Request to generate bulk responses"""
    review_ids: List[int] = Field(..., min_items=1, max_items=50)
    template_id: Optional[int] = None
    auto_send: Optional[bool] = False
    business_name: Optional[str] = None


class BulkResponseResult(BaseModel):
    """Result of bulk response operation"""
    review_id: int
    success: bool
    response_id: Optional[int] = None
    error: Optional[str] = None
    response_text: Optional[str] = None


class BulkResponseResponse(BaseModel):
    """Response after bulk response operation"""
    success: bool
    message: str
    total_processed: int
    successful_responses: int
    failed_responses: int
    results: List[BulkResponseResult] = Field(default_factory=list)


# Auto-response configuration
class AutoResponseConfig(BaseModel):
    """Configuration for automatic review responses"""
    enabled: bool = False
    respond_to_positive: bool = True
    respond_to_neutral: bool = True
    respond_to_negative: bool = True
    min_rating_threshold: float = Field(1.0, ge=1, le=5)
    max_rating_threshold: float = Field(5.0, ge=1, le=5)
    delay_hours: int = Field(2, ge=0, le=168, description="Hours to wait before auto-responding")
    platforms: List[ReviewPlatform] = Field(default_factory=list)
    business_hours_only: bool = False
    weekend_responses: bool = True
    template_selection_strategy: str = Field("smart", description="smart, random, or highest_priority")


class AutoResponseStats(BaseModel):
    """Statistics for auto-response system"""
    total_auto_responses: int = 0
    auto_responses_today: int = 0
    auto_responses_this_week: int = 0
    auto_responses_this_month: int = 0
    success_rate: float = 0.0
    average_response_time_hours: float = 0.0
    most_used_template: Optional[str] = None
    platform_breakdown: Dict[str, int] = Field(default_factory=dict)